compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/intrinsics/KClassJavaPrimitiveTypeProperty.kt

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright 2000-2018 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.backend.jvm.intrinsics
+
+import org.jetbrains.kotlin.backend.jvm.JvmBackendContext
+import org.jetbrains.kotlin.backend.jvm.codegen.BlockInfo
+import org.jetbrains.kotlin.codegen.AsmUtil
+import org.jetbrains.kotlin.codegen.StackValue
+import org.jetbrains.kotlin.descriptors.FunctionDescriptor
+import org.jetbrains.kotlin.ir.expressions.IrClassReference
+import org.jetbrains.kotlin.ir.expressions.IrGetClass
+import org.jetbrains.kotlin.ir.expressions.IrMemberAccessExpression
+import org.jetbrains.kotlin.psi.KtClassLiteralExpression
+import org.jetbrains.kotlin.resolve.BindingContext
+import org.jetbrains.kotlin.resolve.jvm.AsmTypes
+import org.jetbrains.kotlin.resolve.jvm.jvmSignature.JvmMethodSignature
+import org.jetbrains.kotlin.resolve.scopes.receivers.ExpressionReceiver
+import org.jetbrains.kotlin.types.TypeUtils
+import org.jetbrains.org.objectweb.asm.Type
+import org.jetbrains.org.objectweb.asm.commons.InstructionAdapter
+
+class KClassJavaPrimitiveTypeProperty : IntrinsicMethod() {
+    override fun toCallable(expression: IrMemberAccessExpression, signature: JvmMethodSignature, context: JvmBackendContext): IrIntrinsicFunction {
+        return object: IrIntrinsicFunction(expression, signature, context) {
+            override fun invoke(v: InstructionAdapter, codegen: org.jetbrains.kotlin.backend.jvm.codegen.ExpressionCodegen, data: BlockInfo): StackValue {
+                val extensionReceiver = expression.extensionReceiver
+                if (extensionReceiver is IrGetClass) {
+                    val extensionReceiverType = codegen.gen(extensionReceiver.argument, data).type
+                    when {
+                        extensionReceiverType == Type.VOID_TYPE -> {
+                            v.aconst(null)
+                        }
+                        AsmUtil.isPrimitive(extensionReceiverType) -> {
+                            AsmUtil.pop(v, extensionReceiverType)
+                            v.getstatic(AsmUtil.boxType(extensionReceiverType).getInternalName(), "TYPE", "Ljava/lang/Class;");
+                        }
+                        else -> {
+                            AsmUtil.pop(v, extensionReceiverType)
+                            if (AsmUtil.unboxPrimitiveTypeOrNull(extensionReceiverType) != null) {
+                                v.getstatic(extensionReceiverType.getInternalName(), "TYPE", "Ljava/lang/Class;");
+                            } else {
+                                v.aconst(null)
+                            }
+                        }
+                    }
+                } else if (extensionReceiver is IrClassReference) {
+                    assert(extensionReceiver is IrClassReference)
+                    val classType = (extensionReceiver as IrClassReference).classType
```

Comment:
After three checks, I'd be surprised if it wasn't a `IrClassReference`... :) This cast and the assert above are not necessary